src/expr.rs

```diff
@@ -99,11 +100,74 @@ impl Rewrite for ast::Expr {
                               width,
                               offset)
             }
+            ast::Expr_::ExprClosure(capture, ref fn_decl, ref body) => {
+                rewrite_closure(capture, fn_decl, body, self.span, context, width, offset)
+            }
             _ => context.codemap.span_to_snippet(self.span).ok()
         }
     }
 }
 
+fn rewrite_closure(capture: ast::CaptureClause,
+                   fn_decl: &ast::FnDecl,
+                   body: &ast::Block,
+                   span: Span,
+                   context: &RewriteContext,
+                   width: usize,
+                   offset: usize)
+                   -> Option<String> {
+    let mover = if capture == ast::CaptureClause::CaptureByValue {
+        "move "
+    } else {
+        ""
+    };
+
+    // 4 = "|| {".len(), which is overconservative when the closure consists of
+    // a single expression.
+    let argument_budget = try_opt!(width.checked_sub(4 + mover.len()));
+    // 1 = |
+    let argument_offset = offset + mover.len() + 1;
+
+    let arg_items = itemize_list(context.codemap,
+                                 Vec::new(),
+                                 // FIXME(#185): we shouldn't have to clone here
+                                 fn_decl.inputs.iter().cloned(),
+                                 ",",
+                                 "|",
+                                 span_lo_for_arg,
+                                 span_hi_for_arg,
+                                 rewrite_fn_input,
+                                 span_after(span, "|", context.codemap),
+                                 body.span.lo);
+
+    let fmt = ListFormatting {
+        tactic: ListTactic::HorizontalVertical,
+        separator: ",",
+        trailing_separator: SeparatorTactic::Never,
+        indent: argument_offset,
+        h_width: argument_budget,
+        v_width: argument_budget,
+        ends_with_newline: true,
+    };
+
+    let prefix = format!("{}|{}|", mover, write_list(&arg_items, &fmt));
+
+    let body_str = try_opt!(if body.stmts.is_empty() {
+        let expr = body.expr.as_ref().unwrap();
+        // 1 = the separating space between arguments and the body.
+        let extra_offset = extra_offset(&prefix, offset) + 1;
+
+        expr.rewrite(context, width - extra_offset, offset + extra_offset)
```

Comment:
If this fails, we should stick the expression on the next line, rather than just failing ourselves. We should really insert braces too. In fact, we might want to insert braces in any case if the body spans multiple lines (although that would mean reformatting the body again).